### PR TITLE
NXP-26209: handle WOPI discovery refresh

### PR DESCRIPTION
**Pull request by ataillefer**

### CODE REVIEW

File: nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIService.java [hunk 1]
```diff
@@ -51,4 +51,13 @@
     boolean verifyProofKey(String proofKeyHeader, String oldProofKeyHeader, String url, String accessToken,
             String timestampHeader);
 
+    /**
+     * Refreshes the WOPI discovery: fetches the data from the discovery URL, stores it and loads it in memory.
+     * <p>
+     * Returns {@code true} if refreshing the discovery has succeeded, {@code false} otherwise.
```
Comment (efge) [inline nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIService.java]: Probably better in the `@return` section

File: nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIServiceImpl.java [hunk 2]
```diff
@@ -213,4 +242,81 @@ public boolean verifyProofKey(String proofKeyHeader, String oldProofKeyHeader, S
...
+        byte[] discoveryBytes = fetchDiscovery();
+        if (ArrayUtils.isEmpty(discoveryBytes)) {
+            return false;
+        }
+
+        log.debug("Successfully fetched WOPI dicovery");
+        storeDiscovery(discoveryBytes);
+        loadDiscovery(discoveryBytes);
+        if (Framework.isBooleanPropertyTrue(CLUSTERING_ENABLED_PROP)) {
+            invalidator.sendMessage(new WOPIDiscoveryInvalidation(discoveryBytes));
```
Comment (efge) [inline nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIServiceImpl.java]: At this point `CLUSTERING_ENABLED_PROP` is an implementation detail, you should check `invalidator != null` instead.

File: nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIServiceImpl.java [hunk 3]
```diff
@@ -213,4 +242,81 @@ public boolean verifyProofKey(String proofKeyHeader, String oldProofKeyHeader, S
...
+        keyValueStore.put(WOPI_DISCOVERY_KEY, discoveryBytes);
+    }
+
+    public static class WOPIDiscoveryInvalidation implements SerializableMessage {
+
+        private static final long serialVersionUID = 1L;
+
+        public final byte[] discovery;
+
+        public WOPIDiscoveryInvalidation(byte[] discovery) {
```
Comment (efge) [inline nuxeo-wopi/src/main/java/org/nuxeo/wopi/WOPIServiceImpl.java]: But you don't need to pass the discovery bytes in the invalidation message, it's already available in the key/value store.

**Comment by efge** [comment]
👍

**Comment by troger** [comment]
:+1:

**Pull request merged**